hi-ml-histopathology/testhisto/testhisto/datasets/test_panda_slides.py

```diff
@@ -0,0 +1,91 @@
+from pytorch_lightning import seed_everything
+
+from histopathology.configs.classification.DeepSMILESlidesPandaBenchmark import DeepSMILESlidesPandaBenchmark
+from histopathology.utils.naming import SlideKey
+from testhisto.mocks.base_data_generator import MockHistoDataType
+from testhisto.mocks.slides_generator import MockPandaSlidesGenerator, TilesPositioningType
+
+try:
+    from cucim import CuImage  # noqa: F401
+    has_cucim = True
+except:  # noqa: E722
+    has_cucim = False
+
+
+@pytest.mark.skipif(not has_cucim, reason="Test requires CUCIM library")
+@pytest.mark.gpu
+def test_panda_reproducibility(tmp_path: Path) -> None:
+    """Check if subsequent enumerations of the Panda dataset produce identical sequences of tiles."""
+    seed_everything(seed=123, workers=True)
+    tile_size = 28
+    num_tiles = 4
+    wsi_generator = MockPandaSlidesGenerator(
+        dest_data_path=tmp_path,
+        mock_type=MockHistoDataType.FAKE,
+        n_tiles=num_tiles,
+        n_slides=10,
+        n_channels=3,
+        n_levels=3,
+        tile_size=tile_size,
+        background_val=255,
+        tiles_pos_type=TilesPositioningType.RANDOM
+    )
+    wsi_generator.generate_mock_histo_data()
+
+    container = DeepSMILESlidesPandaBenchmark()
+    container.tile_size = tile_size
+    container.max_bag_size = num_tiles
+    container.local_datasets = [tmp_path]
+
+    def test_data_items_are_equal(loader_fn_names: List[str]) -> None:
+        """Creates a new data module from the container, and checks if all the data loaders specified in
+        `loader_fn_names` return the same set of items when enumerated twice."""
+        data_module = container.get_data_module()
+        data_module.dataloader_kwargs = {**data_module.dataloader_kwargs,
+                                         "multiprocessing_context": None,
+                                         "num_workers": 0}
+
+        def get_item_from_new_loader(loader_fn_name: str) -> Any:
+            """Get a new dataloader with the given name (for example, 'train_dataloader') and returns the first item."""
+            loader_fn = getattr(data_module, loader_fn_name)
+            loader = loader_fn()
+            iterator = iter(loader)
+            return next(iterator)
+
+        for loader_fn_name in loader_fn_names:
+            print(f"Checking {loader_fn_name}")
+            item1 = get_item_from_new_loader(loader_fn_name)
+            item2 = get_item_from_new_loader(loader_fn_name)
+            assert item1[SlideKey.SLIDE_ID] == item2[SlideKey.SLIDE_ID], "Order of slides must match"
+            assert len(item1[SlideKey.IMAGE]) == len(item2[SlideKey.IMAGE]), "Length of images must match"
+            for i in range(len(item1[SlideKey.IMAGE])):
+                image1 = item1[SlideKey.IMAGE][i]
+                image2 = item2[SlideKey.IMAGE][i]
+                if len(np.unique(image1)) == 1:
+                    assert False, "Something is wrong here, image1 only has a single value"
```

Comment:
Is it really something wrong? a tile can be complete background and therefore have a single value. It might be because max_bag_size is bigger than the effective foreground in the WSI, in which case, we should rather make sure that we get a background tile deterministically. torch.allclose(image1, image2) should be valid in this case as well.
This is a special case where the mock WSI have exactly num_tiles of foreground that we use for tiling as well.
But in real life, not all slides will have the expected num_tiles and might include some background tiles.

Reply:
I added that check specifically because with all-ones slides the check below is meaningless. I agree that under normal circumstances it could happen that a tile is only background. Here in this test we are using a fixed seed, and will for sure not run into that problem.